README.pod
=pod

=head1 NAME

App::RewriteVersion - A tool to rewrite and bump your Perl module versions

=head1 SYNOPSIS

 use App::RewriteVersion;
 my $app = App::RewriteVersion->new;
 
 # Options
 $app->verbose(1)->follow_symlinks(0);
 
 # Bump versions for modules in current dist directory
 $app->rewrite_versions($app->bump_version($app->current_version));
 
 # Bump versions in specified dist directory
 $app->rewrite_versions($app->bump_version($app->current_version(dist_dir => $dist_dir)), dist_dir => $dist_dir);
 
 # Override module to read version from
 $app->rewrite_versions($app->bump_version($app->current_version(file => $file)));
 
 # Bump versions in specific subdirectories
 $app->rewrite_versions($app->bump_version($app->current_version), subdirs => ['foo','bar']);
 
 # Custom version bump algorithm
 $app->rewrite_versions($app->bump_version($app->current_version, sub { shift + 0.05 }));
 
 # Don't bump, just synchronize versions with main module
 $app->rewrite_versions($app->current_version);
 
 # Set versions to specified version
 $app->rewrite_versions('0.065');
 
=head1 DESCRIPTION

L<App::RewriteVersion> is a tool for managing Perl module versions in a
distribution. It is heavily based on the L<Dist::Zilla> plugin
L<Dist::Zilla::Plugin::RewriteVersion>. Similarly to that plugin, the C<V>
environment variable can be used to override the version detected from the main
module.

Existing version assignments and new versions must be parseable with the same
rules as in L<Dist::Zilla::Plugin::RewriteVersion/"DESCRIPTION">, that is to
say, they should either be a decimal number with a single decimal point, or a
tuple version with a leading C<v> and at least 3 segments separated by decimal
points. Version assignments should be in the form C<our $VERSION = '...';>.

See L<perl-rewrite-version> and L<perl-bump-version> for details on
command-line usage.

=head1 ATTRIBUTES

=head2 allow_decimal_underscore

 my $bool = $app->allow_decimal_underscore;
 $app = $app->allow_decimal_underscore(0);

If true, decimal versions with underscores will be allowed. Defaults to true.
See L<Dist::Zilla::Plugin::BumpVersionAfterRelease/"Using underscore in decimal $VERSION">
for more information.

=head2 dry_run

 my $bool = $app->dry_run;
 $app = $app->dry_run(1);

If true, the module will process files as normal but not actually modify them.
Useful with L</"verbose"> to verify expected functionality.

=head2 follow_symlinks

 my $bool = $app->follow_symlinks;
 $app = $app->follow_symlinks(1);

If true, the application will follow symlinked directories when traversing the
distribution for modules. Defaults to false.

=head2 global

 my $bool = $app->global;
 $app = $app->global(1);

If true, the application will replace all version assignments found instead of
just the first instance in each file. Defaults to false.

=head2 verbose

 my $bool = $app->verbose;
 $app = $app->verbose(1);

Enable progress messages to be printed to STDOUT. Defaults to false.

=head1 METHODS

=head2 new

 my $app = App::RewriteVersion->new;

Construct a new L<App::RewriteVersion> object.

=head2 bump_version

 my $new_version = $app->bump_version($version);
 my $new_version = $app->bump_version($version, sub { $_[0] + 1 });

Increments a version string, returning the new version string. An optional
coderef can be passed for custom version bump logic. The coderef will receive
the current version string as the first argument and is expected to return a
new version string. By default, L<Version::Next> is used. An exception will be
thrown if an invalid version is passed according to the current settings.

=head2 current_version

 my $current_version = $app->current_version;
 my $current_version = $app->current_version(dist_dir => $dist_dir);
 my $current_version = $app->current_version(file => $file);

Returns the current version of the distribution using L</"version_from">. If no
C<file> is passed, the main module filename will be guessed from C<dist_dir>
(defaulting to current working directory), using heuristics similar to
L<Dist::Zilla::Plugin::NameFromDirectory> and L<Dist::Zilla/"main_module">. For
example, if the directory is named C<Foo-Bar> it will look for
C<lib/Foo/Bar.pm> within the distribution, and if that doesn't exist, the
shortest module path found. If the C<V> environment variable is set, it will be
returned regardless of other options. An exception will be thrown if no version
is found in the specified or guessed file, or if no perl modules could be
found.

=head2 rewrite_version

 my $bool = $app->rewrite_version($file, $version);
 my $bool = $app->rewrite_version($file, $version, is_trial => $is_trial);

Rewrites the version of the file at C<$file> to C<$version> if it has a version
assignment in the form C<our $VERSION = '...';>. Returns true if the version
was rewritten, or false if no version assignment was found. If C<is_trial> is
true, C<# TRIAL> will be appended to the version assignment line when
rewriting. The C<-TRIAL> indication is not part of the version and should be
added to the name of the archive you upload to PAUSE, which is outside of the
scope of this tool. An exception will be thrown if an invalid version is
passed, or an I/O error occurs. A warning will be printed if C<$version> is
lower than the version previously declared in the file.

=head2 rewrite_versions

 $app = $app->rewrite_versions($version);
 $app = $app->rewrite_versions($version, dist_dir => $dist_dir);
 $app = $app->rewrite_versions($version, is_trial => 1);
 $app = $app->rewrite_versions($version, subdirs => ['lib']);

Rewrites the versions of all perl files found in C<dist_dir> (defaulting to
current working directory) to C<$version> using L</"rewrite_version">. The
C<subdirs> option can be used to specify an arrayref of subdirectories relative
to C<dist_dir> in which versions will be rewritten, otherwise defaulting to
C<lib>, C<script>, and C<bin>. If passed, the C<is_trial> option is passed
through to L</"rewrite_version">. An exception will be thrown if an invalid
version is passed, or an I/O error occurs.

=head2 version_from

 my $version = $app->version_from($file);

Attempts to read version from the file at C<$file>. Returns C<undef> if no
version assignment was found.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2015 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Dist::Zilla::Plugin::RewriteVersion>, L<Version::Next>

=cut


Changes
{{$NEXT}}

1.000     2018-12-13 14:06:09 EST
  - Default allow_decimal_underscore to enabled. Underscores in decimal
    versions are not ideal, but they are still the most convenient method of
    indicating a developer release when not using Dist::Zilla.

0.006     2018-04-21 19:40:09 EDT
  - Use and recognize tr/_//d to remove underscores from versions

0.005     2016-09-10 16:33:08 EDT
  - Lower perl version requirement to 5.8.1
  - Move Unicode::UTF8 to recommends

0.004     2015-10-30 12:38:37 EDT
  - Use Class::Tiny::Chained

0.003     2015-10-11 03:18:15 EDT
  - Fix loose version check so that version tuples with underscores are
    correctly disallowed

0.002     2015-09-12 20:10:22 EDT
  - Use Getopt::Long::Modern
  - Add warning when version is rewritten to a lower version

0.001     2015-09-02 18:38:57 EDT
  - First release


Build.PL
# This Build.PL for App-RewriteVersion was generated by
# Dist::Zilla::Plugin::ModuleBuildTiny::Fallback 0.025
use strict;
use warnings;

my %configure_requires = (
    'Module::Build::Tiny' => '0.034',
);

my %errors = map {
    eval "require $_; $_->VERSION($configure_requires{$_}); 1";
    $_ => $@,
} keys %configure_requires;

if (!grep { $_ } values %errors)
{
    # This section for App-RewriteVersion was generated by Dist::Zilla::Plugin::ModuleBuildTiny 0.015.
    use strict;
    use warnings;

    use 5.008001;
    # use Module::Build::Tiny 0.034;
    Module::Build::Tiny::Build_PL();
}
else
{
    if (not $ENV{PERL_MB_FALLBACK_SILENCE_WARNING})
    {
        warn <<'EOW'
*** WARNING WARNING WARNING WARNING WARNING WARNING WARNING WARNING ***

If you're seeing this warning, your toolchain is really, really old* and you'll
almost certainly have problems installing CPAN modules from this century. But
never fear, dear user, for we have the technology to fix this!

If you're using CPAN.pm to install things, then you can upgrade it using:

    cpan CPAN

If you're using CPANPLUS to install things, then you can upgrade it using:

    cpanp CPANPLUS

If you're using cpanminus, you shouldn't be seeing this message in the first
place, so please file an issue on github.

This public service announcement was brought to you by the Perl Toolchain
Gang, the irc.perl.org #toolchain IRC channel, and the number 42.

----

* Alternatively, you are running this file manually, in which case you need
to learn to first fulfill all configure requires prerequisites listed in
META.yml or META.json -- or use a cpan client to install this distribution.

You can also silence this warning for future installations by setting the
PERL_MB_FALLBACK_SILENCE_WARNING environment variable, but please don't do
that until you fix your toolchain as described above.


Errors from configure prereqs:
EOW
        . do {
            require Data::Dumper; Data::Dumper->new([ \%errors ])->Indent(2)->Terse(1)->Sortkeys(1)->Dump;
        };

        sleep 10 if -t STDIN && (-t STDOUT || !(-f STDOUT || -c STDOUT));
    }


    # This section was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.009.
    use strict;
    use warnings;

    require Module::Build; Module::Build->VERSION(0.28);


    my %module_build_args = (
      "configure_requires" => {
        "Module::Build::Tiny" => "0.034"
      },
      "dist_abstract" => "A tool to rewrite and bump your Perl module versions",
      "dist_author" => [
        "Dan Book <dbook\@cpan.org>"
      ],
      "dist_name" => "App-RewriteVersion",
      "dist_version" => "0.006",
      "license" => "artistic_2",
      "module_name" => "App::RewriteVersion",
      "recommends" => {
        "Unicode::UTF8" => "0.58"
      },
      "recursive_test_files" => 1,
      "requires" => {
        "Carp" => 0,
        "Class::Tiny::Chained" => 0,
        "Getopt::Long::Modern" => 0,
        "Path::Iterator::Rule" => 0,
        "Path::Tiny" => "0.060",
        "Version::Next" => 0,
        "perl" => "5.008001",
        "version" => "0.77"
      },
      "script_files" => [
        "script/perl-bump-version",
        "script/perl-rewrite-version"
      ],
      "test_requires" => {
        "File::Spec" => 0,
        "File::Temp" => 0,
        "Module::Metadata" => 0,
        "Test::More" => "0.88"
      }
    );


    my %fallback_build_requires = (
      "File::Spec" => 0,
      "File::Temp" => 0,
      "Module::Metadata" => 0,
      "Test::More" => "0.88"
    );


    unless ( eval { Module::Build->VERSION(0.4004) } ) {
      delete $module_build_args{test_requires};
      $module_build_args{build_requires} = \%fallback_build_requires;
    }

    my $build = Module::Build->new(%module_build_args);


    $build->create_build_script;
}


lib/App/BumpVersion.pm
package App::BumpVersion;

use strict;
use warnings;
use Carp 'croak';
use Path::Iterator::Rule;
use Path::Tiny;
use Version::Next 'next_version';

our $VERSION = '0.001';

sub new { bless {}, shift }

sub allow_decimal_underscore { $self->_boolean('allow_decimal_underscore', 0, @_) }

sub follow_symlinks { $self->_boolean('follow_symlinks', 0, @_) }

sub global { $self->_boolean('global', 0, @_) }

sub verbose { $self->_boolean('verbose', 0, @_) }

sub _boolean {
	my ($self, $attr, $default) = (shift, shift, shift);
	return $self->{$attr} //= $default unless @_;
	$self->{$attr} = shift ? 1 : 0;
	return $self;
}

sub bump_versions {
	my $self = shift;
	my %params = @_;
	my $dir = path($params{dir} // '.');
	my $bump = $params{bump} // 1;
	my $trial = $params{is_trial} ? 1 : 0;
	my $version_from = $params{version_from};
	
	my $version;
	if (exists $ENV{V}) {
		$version = $ENV{V};
	} else {
		$version_from //= $self->_main_module($dir);
		$version = $self->version_from($version_from)
			// croak qq{No version found in module "$main_module"};
		$self->_check_version($version);
		if (ref $bump eq 'CODE') {
			$version = $bump->($version);
		} elsif ($bump) {
			$version = next_version($version);
		}
	}
	
	$self->rewrite_versions($dir, $version, $trial);
	
	return $self;
}

sub rewrite_version {
	my ($self, $path, $version, $trial) = @_;
	
	return 0 unless -T $path;
	$path = path($path);
	my $content = $path->slurp_utf8;
	
	my $code = qq{our \$VERSION = '$version';};
	$code .= " # TRIAL" if $trial;
	
	$code .= qq{\n\$VERSION = eval \$VERSION;}
		if $version =~ /_/ and scalar($version =~ /\./g) <= 1;
	
	my $assign_regex = _assign_re();
	if ($self->global ? ($content =~ s{^$assign_regex[^\n]*$}{$code}msg)
	                  : ($content =~ s{^$assign_regex[^\n]*$}{$code}ms)) {
		$path->append_utf8({truncate => 1}, $content);
		return 1;
	}
	
	return 0;
}

sub rewrite_versions {
	my ($self, $dir, $version, $trial) = @_;
	
	$self->_check_version($version);
	
	my $rule = Path::Iterator::Rule->new->file->ascii->skip_vcs->perl_file;
	my %options = (follow_symlinks => $self->follow_symlinks);
	my $iter = $rule->iter("$dir", \%options);
	while (defined(my $file = $iter->())) {
		my $rewritten = $self->rewrite_version($file, $version, $trial);
		if ($self->verbose) {
			print $rewritten ? qq{Updated \$VERSION assignment in "$file"\n}
				: qq{Skipping: no "our \$VERSION = '...'" found in "$file"\n};
		}
	}
	return $self;
}

sub version_from {
	my ($self, $path) = @_;
	
	return undef unless -T $path;
	$path = path($path);
	my $content = $path->slurp_utf8;
	
	my $assign_regex = _assign_re();
	my ($quote, $version) = $content =~ m{^$assign_regex[^\n]*$}ms;
	
	print qq{Extracted version from $path: $version\n} if $version and $self->verbose;
	return $version;
}

sub _check_version {
	my ($self, $version) = @_;
	croak qq{$version is not an allowed version string} unless
		$self->allow_decimal_underscore ? _is_loose_version($version) : _is_strict_version($version);
	return $self;
}

sub _dist_name {
	my ($self, $path) = @_;
	
	# Adapted from Dist::Zilla::Plugin::NameFromDirectory
	my $name = $path->absolute->basename;
	$name =~ s/(?:^(?:perl|p5)-|[\-\.]pm$)//;
	print qq{Guessing distribution name is $name} if $self->verbose;
	
	return $name;
}

sub _main_module {
	my ($self, $path, $name) = @_;
	
	# Adapted from Dist::Zilla
	my $main;
	if (defined $self->main_module) {
		$main = path($self->main_module);
		croak qq{Specified main module at "$main" does not exist} unless $main->exists;
	} else {
		(my $guess = $self->_dist_name($path)) =~ s{-}{/}g;
		$main = $path->child("lib/$guess");
		unless ($main->exists) {
			$main = path($self->_shortest_module($path));
		}
		croak qq{Could not find any modules to retrieve version from}
			unless defined $main and $main->exists;
	}
	
	print qq{Using "$main" as dist's main module\n} if $self->verbose;
	return $main;
}

sub _shortest_module {
	my ($self, $path) = @_;
	$path = $path->child('lib');
	my $rule = Path::Iterator::Rule->new->file->ascii->skip_vcs->perl_module;
	my %options = (follow_symlinks => $self->follow_symlinks);
	return (sort { length $a <=> length $b } $rule->all($path, \%options))[0];
}

# this section copied from Dist::Zilla::Plugin::BumpVersionAfterRelease::_Util
{

# version regexes from version.pm
my $FRACTION_PART              = qr/\.[0-9]+/;
my $STRICT_INTEGER_PART        = qr/0|[1-9][0-9]*/;
my $LAX_INTEGER_PART           = qr/[0-9]+/;
my $STRICT_DOTTED_DECIMAL_PART = qr/\.[0-9]{1,3}/;
my $LAX_DOTTED_DECIMAL_PART    = qr/\.[0-9]+/;
my $LAX_ALPHA_PART             = qr/_[0-9]+/;
my $STRICT_DECIMAL_VERSION     = qr/ $STRICT_INTEGER_PART $FRACTION_PART? /x;
my $STRICT_DOTTED_DECIMAL_VERSION =
  qr/ v $STRICT_INTEGER_PART $STRICT_DOTTED_DECIMAL_PART{2,} /x;
my $STRICT = qr/ $STRICT_DECIMAL_VERSION | $STRICT_DOTTED_DECIMAL_VERSION /x;
my $LAX_DECIMAL_VERSION =
  qr/ $LAX_INTEGER_PART (?: \. | $FRACTION_PART $LAX_ALPHA_PART? )?
    |
    $FRACTION_PART $LAX_ALPHA_PART?
    /x;
my $LAX_DOTTED_DECIMAL_VERSION = qr/
    v $LAX_INTEGER_PART (?: $LAX_DOTTED_DECIMAL_PART+ $LAX_ALPHA_PART? )?
    |
    $LAX_INTEGER_PART? $LAX_DOTTED_DECIMAL_PART{2,} $LAX_ALPHA_PART?
    /x;

sub _is_strict_version { defined $_[0] && $_[0] =~ qr/\A $STRICT \z /x }

sub _is_loose_version {
    defined $_[0] && $_[0] =~ qr/\A $STRICT | $LAX_DECIMAL_VERSION \z /x;
}

# Because this is used for *capturing* or *replacing*, we take anything
# that is a lax version (but not literal string 'undef', so we don't want
# version::LAX).  Later anything captured needs to be checked with the
# strict or loose version check functions.
sub _assign_re {
    return qr{
        our \s+ \$VERSION \s* = \s*
        (['"])($LAX_DECIMAL_VERSION | $LAX_DOTTED_DECIMAL_VERSION)\1 \s* ;
        (?:\s* \# \s TRIAL)? [^\n]*
        (?:\n \$VERSION \s = \s eval \s \$VERSION;)?
    }x;
}

}
# end of copied section

1;

=head1 NAME

App::BumpVersion - A tool to rewrite or bump your Perl module versions

=head1 SYNOPSIS

 use App::BumpVersion;
 my $app = App::BumpVersion->new;
 
 # Options
 $app->verbose(1)->follow_symlinks(0);
 
 # Bump versions for modules in current dist directory
 $app->bump_versions;
 
 # Bump versions in specified dist directory
 $app->bump_versions(dir => 'Foo-Bar/');
 
 # Override module to read version from
 $app->bump_versions(version_from => 'lib/Foo/Bar.pm');
 
 # Don't bump, just synchronize versions with main module
 $app->bump_versions(bump => 0);
 
 # Set versions to specified version (or use the V environment variable)
 $app->bump_versions(bump => sub { '0.56' });
 
 # Custom version bump algorithm
 $app->bump_versions(bump => sub { shift + 0.05 });
 
=head1 DESCRIPTION

L<App::BumpVersion> is a tool for managing Perl module versions in a
distribution. It is heavily based on the L<Dist::Zilla> plugin
L<Dist::Zilla::Plugin::RewriteVersion>. Similarly to that plugin, the C<V>
environment variable can be used to override the version that is set, but note
that in this case, it overrides the version-bumping functionality altogether.
Existing version assignments and new versions must be parseable with the same
rules as in L<Dist::Zilla::Plugin::RewriteVersion/"DESCRIPTION">.

See L<perl-bump-version> for details on command-line usage.

=head1 ATTRIBUTES

=head2 allow_decimal_underscore

 my $bool = $app->allow_decimal_underscore
 $app = $app->allow_decimal_underscore(1);

If true, decimal versions with underscores will be allowed. Defaults to false.
As with L<Dist::Zilla::Plugin::RewriteVersion>, version tuples with underscores
are never allowed.

=head2 follow_symlinks

 my $bool = $app->follow_symlinks;
 $app = $app->follow_symlinks(1);

If true, the application will follow symlinked directories when traversing the
distribution for modules. Defaults to false.

=head2 global

 my $bool = $app->global;
 $app = $app->global(1);

If true, the application will replace all version assignments found in each
particular module instead of just the first instance found. Defaults to false.

=head2 verbose

 my $bool = $app->verbose;
 $app = $app->verbose(1);

Enable progress messages to be printed to STDOUT. Defaults to false.

=head1 METHODS

=head2 new

 my $app = App::BumpVersion->new;

Construct a new L<App::BumpVersion> object.

=head2 bump_versions

 $app = $app->bump_versions;
 $app = $app->bump_versions(dir => '/foo/projects/Acme-Bar');
 $app = $app->bump_versions(bump => sub { 'v5.0.1' });
 $app = $app->bump_versions(version_from => 'lib/My/Module.pm');
 $app = $app->bump_versions(is_trial => 1);

Rewrites versions in all perl files found that contain a version assignment in
the form C<our $VERSION = '...'>. See L<Dist::Zilla::Plugin::RewriteVersion>
for more information on the version parsing semantics. Accepts the following
options:

=over

=item * bump

If set to a code reference, it will be called with the current version as a
string, and will be expected to return the new version to use for rewriting.
Otherwise, a true value will use L<Version::Next> to determine the version to
set, and a false value will use the original unchanged version. Defaults to
true.

=item * dir

Relative or absolute path for the distribution directory to operate on.
Defaults to the current working directory.

=item * is_trial

If true, C<# TRIAL> will be appended to the version assignment line. Defaults
to false.

=item * version_from

File to read version from before bumping (if applicable), relative to the
distribution root directory. If unset, the main module filename will be guessed
from the distribution directory name, using heuristics similar to
L<Dist::Zilla::Plugin::NameFromDirectory> and L<Dist::Zilla/"main_module">.

=back

=head2 rewrite_version

 my $bool = $app->rewrite_version($path, $version, $is_trial);

Rewrites the version of the file at C<$path> to C<$version>. C<$path> may be a
L<Path::Tiny> object. Returns true if the version was rewritten, or false if no
version assignment was found.

=head2 rewrite_versions

 $app = $app->rewrite_versions($dir, $version, $is_trial);

Rewrites the versions of all perl files found in C<$dir> to C<$version> using
L</"rewrite_version">.

=head2 version_from

 my $version = $app->version_from($path);

Attempts to read version from the file at C<$path>. Returns C<undef> if no
version assignment was found.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2015 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Dist::Zilla::Plugin::RewriteVersion>, L<Version::Next>
